ministryofjustice/analytics-platform-ops
# Add kubernetes cluster Terraform module, moving all cluster creation and almost all cluster management tasks into Terraform

## What

**NOTE: this PR should not be merged just yet until some changes have been made to ensure that it does not cause issues with existing environments**

This PR adds Terraform modules to handle creation and management of all AWS resources for a Kubernetes cluster. This is based on the terraform output of [kops](https://github.com/kubernetes/kops), which has been converted to a parameterised terraform module.

This Terraform module also allows for easy deployment of a development environment within AWS with minimal resources - this consists of:
- 1x SSH bastion host (t2.micro)
- 1x kubernetes master (t2.small)
- 1x kubernetes node (t2.small)
- kubernetes API ELB

The `kops` binary is also wrapped in Terraform `local-exec` resources, meaning _most_ cluster management tasks (such as Kubernetes version upgrades) will be handled automatically during a `terraform apply` run. However not all cluster changes will be automatically applied during an `apply` due to the dependence on `kops`, which has some issues related to detecting a need for updates - for instance, changing `kubeAPIServer` flags in kops `cluster.tpl.yml` will not automatically trigger a cluster update, and will require `kops rolling-update cluster --force` after running `terraform apply`.

## How to review

1. Check out this branch
2. `$ cd infra/terraform/environments` 
3. `$ cp dev-env-example $DEV_ENV_NAME`
4. `$ cd $DEV_ENV_NAME`
5. Edit `main.tf` and replace `dev-env-example` in `key = "dev-env-example/terraform.tfstate"` in the `terraform` block at the top of the file with your env name.
6. Edit `terraform.tfvars` (or copy `terraform.example.tfvars` to `terraform.tfvars` if you are not in this repo's git-crypt keychain) and specify your env name in `env = ""`, and enter an SSH public key in `ssh_public_key = ""`
7. `$ terraform init`
8. `$ terraform plan`
9. `$ terraform apply`

Initial environment creation can take 20+ minutes due to the need to wait for new DNS records to resolve, and for Kubernetes components to self-configure and report themselves as healthy.

Once `terraform apply` has completed your local `kubectl` environment should be configured for this environment and ready to go - confirm that this is the case by either running `kubectl cluster-info` or `kops validate cluster`.
